pkg/node/manager/manager.go

```diff
@@ -332,134 +333,138 @@ func (m *manager) legacyNodeIpBehavior() bool {
+	// through the tunnel to preserve the source identity as part of the
+	// encapsulation. In encryption case we also want to use vxlan device
+	// to create symmetric traffic when sending nodeIP->pod and pod->nodeIP.
+	return address.Type == addressing.NodeCiliumInternalIP || m.conf.EncryptionEnabled() ||
+		option.Config.EnableHostFirewall || option.Config.JoinCluster
+}
+
+func (m *manager) nodeAddressHasEncryptKey(address nodeTypes.Address) bool {
+	return (m.conf.NodeEncryptionEnabled() ||
+		// If we are doing encryption, but not node based encryption, then do not
+		// add a key to the nodeIPs so that we avoid a trip through stack and attempting
+		// to encrypt something we know does not have an encryption policy installed
+		// in the datapath. By setting key=0 and tunnelIP this will result in traffic
+		// being sent unencrypted over overlay device.
+		(address.Type != addressing.NodeExternalIP && address.Type != addressing.NodeInternalIP)) &&
+		// Also ignore any remote node's key if the local node opted to not perform
+		// node-to-node encryption
+		!node.GetOptOutNodeEncryption()
+}
+
+func (m *manager) nodeAddressSkipsIPCache(address nodeTypes.Address) bool {
+	return m.legacyNodeIpBehavior() && address.Type != addressing.NodeCiliumInternalIP
+}
+
+func nodeIdentityLabels(n nodeTypes.Node) (nodeLabels labels.Labels, hasOverride bool) {
+	nodeLabels = labels.NewFrom(labels.LabelRemoteNode)
+	if n.IsLocal() {
+		nodeLabels = labels.NewFrom(labels.LabelHost)
+	} else if !identity.NumericIdentity(n.NodeIdentity).IsReservedIdentity() {
+		// This needs to match clustermesh-apiserver's VMManager.AllocateNodeIdentity
+		nodeLabels = labels.Map2Labels(n.Labels, labels.LabelSourceK8s)
+		hasOverride = true
+	}
+	return nodeLabels, hasOverride
+}
+
 // NodeUpdated is called after the information of a node has been updated. The
 // node in the manager is added or updated if the source is allowed to update
 // the node. If an update or addition has occurred, NodeUpdate() of the datapath
 // interface is invoked.
 func (m *manager) NodeUpdated(n nodeTypes.Node) {
 	log.Debugf("Received node update event from %s: %#v", n.Source, n)
 
-	nodeIdentity := n.Identity()
+	nodeIdentifier := n.Identity()
 	dpUpdate := true
-	nodeIP := n.GetNodeIP(false)
-
-	remoteHostIdentity := identity.ReservedIdentityHost
-	if m.conf.RemoteNodeIdentitiesEnabled() {
-		nid := identity.NumericIdentity(n.NodeIdentity)
-		if nid != identity.IdentityUnknown && nid != identity.ReservedIdentityHost {
-			remoteHostIdentity = nid
-		} else if !n.IsLocal() {
-			remoteHostIdentity = identity.ReservedIdentityRemoteNode
-		}
+	var nodeIP netip.Addr
+	if nIP := n.GetNodeIP(false); nIP != nil {
+		// Skip returning the error here because at this level, we assume that
+		// the IP is valid as long as it's coming from nodeTypes.Node. This
+		// object is created either from the node discovery (K8s) or from an
+		// event from the kvstore.
+		nodeIP, _ = ip.AddrFromIP(nIP)
```

Comment:
Maybe worth filing an issue to support IPv6-only nodes and add `GH-XXXX:` here as a reference?

Reply:
https://github.com/cilium/cilium/issues/24829